src/dbnode/persist/fs/seek_manager.go

```diff
@@ -286,15 +343,188 @@ func (m *seekerManager) Return(shard uint32, start time.Time, seeker ConcurrentD
+	// goroutine will need to wait for the other goroutine to finish before proceeding. The check is performed in
+	// a loop because each iteration relinquishes the lock temporarily. Once the lock is reacquired the same
+	// conditions need to be checked again until this Goroutine finds that either:
+	//
+	//   a) Seekers are already present for this blockStart in which case it can proceed to hot swap them.
+	// or
+	//   b) Seeks are not present for this blockStart and no other goroutines are currently trying to open them,
+	//      in which case this goroutine can just set the new seekers and be done.
+	for {
+		byTime.Lock()
+		seekers, ok = byTime.seekers[blockStartNano]
+
+		if !ok || seekers.active.wg == nil {
+			// Exit the loop still holding the lock.
+			break
+		}
+
+		// If another goroutine is currently trying to open seekers for this block start
+		// then wait for that operation to complete.
+		wg := seekers.active.wg
+		byTime.Unlock()
+		wg.Wait()
+	}
+
+	if !ok {
+		// No existing seekers, so just set the newly created ones and be done.
+		seekers.active = newActiveSeekers
+		if seekers.active.volume > state.Volume {
+			// Ignore any close errors because its not relevant from the callers perspective.
+			m.closeSeekersAndLogError(descriptor, newActiveSeekers.seekers)
+			byTime.Unlock()
+			return 0, errOutOfOrderUpdateOpenLease
+		}
+
+		byTime.seekers[blockStartNano] = seekers
+		byTime.Unlock()
+		return updateOpenLeaseResult, nil
+	}
+
+	// Existing seekers exist.
+	updateOpenLeaseResult = block.UpdateOpenLease
+
+	if seekers.active.volume > state.Volume {
+		// Ignore any close errors because its not relevant from the callers perspective.
+		m.closeSeekersAndLogError(descriptor, newActiveSeekers.seekers)
+		byTime.Unlock()
+		// TODO(rartoul): Should this just be an error?
+		return 0, errOutOfOrderUpdateOpenLease
+	}
+
+	seekers.inactive = seekers.active
+	seekers.active = newActiveSeekers
+
+	// If any of the previous seekers are still borrowed this function will need to wait for
+	// them to be returned.
+	anySeekersAreBorrowed := false
+	for _, seeker := range seekers.inactive.seekers {
+		if seeker.isBorrowed {
+			anySeekersAreBorrowed = true
+			break
+		}
+	}
+
+	var wg *sync.WaitGroup
+	if anySeekersAreBorrowed {
+		// If any of the seekers are borrowed setup a waitgroup which will be used to
+		// signal when they've all been returned (the last seeker that is returned via
+		// the Return() API will call wg.Done()).
+		wg = &sync.WaitGroup{}
+		wg.Add(1)
+		seekers.inactive.wg = wg
+	}
+	byTime.seekers[blockStartNano] = seekers
+	byTime.Unlock()
+
+	if wg != nil {
+		// Wait for all the inactive seekers to be returned and closed because the contract
+		// of this API is that the Leaser (SeekerManager) should have relinquished any resources
+		// associated with the old lease by the time this function returns.
+		wg.Wait()
+	}
+
+	return updateOpenLeaseResult, nil
```

Comment:
I think the above code is hairy enough we should try to break it into some subfunctions so that the high level code can more straight forward manage the locks and waitgroup creation while the subfunctions handle the minutia of each task, what do you think?

Reply:
yeah I think thats fair, let me see what I can do

Reply:
Yeah ok I think I see a way to do it, I'll ping you once I've implemented